### PR TITLE
Add the Isle of Bute Discovery Centre Cinema

### PR DESCRIPTION
**Pull request by andrewferguson**
[This cinema](http://discoverycentrecinema.blogspot.com/) is a small independent (one screen, 97 seats) and the listings are on a Blogspot blog that is updated each month with a hand-edited HTML table.

The "lxml" package has been added to the archiver to support the correct parsing of XML Atom/RSS feeds in BeautifulSoup, such as those used by Blogspot.

### CODE REVIEW

File: src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        row_datas[3] == 'Time' and \
+        row_datas[4] == 'Film - Evening' and \
+        row_datas[5] == 'Time'
+    if not schema_check:
+        raise CinemaArchiverException('Table schema does not match expected format')
+
+
+def parse_table_row(table_row: BeautifulSoup) -> [Showing]:
+    row_datas = [td.text for td in table_row.find_all('td')]
+    showings = [ extract_showing(row_datas, 0), extract_showing(row_datas, 1)]
```
Comment (davidferguson) [inline src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py]:
```suggestion
    showings = [extract_showing(row_datas, 0), extract_showing(row_datas, 1)]
```
Remove the whitespace here please

File: src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    def get_showings(self) -> [Showing]:
+        showings = []
+        feed_contents = get_response(FEED_URL).text
+        feed_soup = BeautifulSoup(feed_contents, features='xml')
+        html_contents = feed_soup.find('entry').find('content', type='html').contents
+        html_soup = BeautifulSoup(html_contents[0], features='html.parser')
+        listings_table = html_soup.find('table')
+        for i, row in enumerate(listings_table.find_all('tr')):
+            if i == 0:
+                continue # skip the 'Film listings for <Month> row'
```
Comment (davidferguson) [inline src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py]:
```suggestion
                continue  # skip the 'Film listings for <Month> row'
```
Two spaces before the comment please

File: src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        row_datas[4] == 'Film - Evening' and \
+        row_datas[5] == 'Time'
+    if not schema_check:
+        raise CinemaArchiverException('Table schema does not match expected format')
+
+
+def parse_table_row(table_row: BeautifulSoup) -> [Showing]:
+    row_datas = [td.text for td in table_row.find_all('td')]
+    showings = [ extract_showing(row_datas, 0), extract_showing(row_datas, 1)]
+    showings = [showing for showing in showings if showing != None]
```
Comment (davidferguson) [inline src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py]:
```suggestion
    showings = [showing for showing in showings if showing is not None]
```

Should be `showing is not None`

File: src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    return showings
+
+
+def extract_showing(row_datas: [str], showing_number: int) -> Showing:
```
Comment (davidferguson) [inline src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py]:
```suggestion
def extract_showing(row_datas: [str], showing_number: int) -> Optional[Showing]:
```

The type should be `Optional[Showing]`.

File: src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
+import re, lxml, datetime
```
Comment (davidferguson) [inline src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py]: Please have one import per line. And `lxml` is not used in this file.

File: setup.py [hunk 6]
```diff
@@ -23,5 +23,6 @@
         'requests',
         'beautifulsoup4',
         'selenium',
+        'lxml',
```
Comment (davidferguson) [inline setup.py:26]: Where is `lxml` used?

**Review by davidferguson — COMMENTED** [review]
A few minor formatting changes, but otherwise looks good.

File: src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py (hunk 5 above)
Comment (andrewferguson) [reply src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py]: Changed to one import per line. And `lxml` is used for parsing the RSS feed using BeautifulSoup:
```
feed_soup = BeautifulSoup(feed_contents, features='xml')
```

File: setup.py (hunk 6 above)
Comment (andrewferguson) [reply setup.py:26]: `lxml` is used for parsing the RSS feed using BeautifulSoup:
```
feed_soup = BeautifulSoup(feed_contents, features='xml')
```

**Comment by andrewferguson** [comment]
The issues highlighted have been addressed, and a bug caused by the hand-edited HTML of this cinema's listings has been corrected.

It runs correctly, and is now ready for review.

**Review by davidferguson — APPROVED** [review]

**Pull request merged**